### ISSUE TITLE
Improve Timeout on Client Agent with CPU configuration

### ISSUE THREAD

**Issue by jhwgh1968** [TabbyML/tabby#689]
I feel like this is a bug, technically speaking? But as noted in other issues, there is not much support for CPU configuration anymore. This is especially since the removal of the < 1B parameter models. So it is probably an enhancement.

While I have gotten everything working, my issue is that the agent has a hard-coded 4 seconds expectation for the average response time. After 3 requests, considers the server unhealthy and then locks out further completions:

https://github.com/TabbyML/tabby/blob/90e446bfbaef76aedd43d46e74c0ba8a92109acb/clients/tabby-agent/src/CompletionProviderStats.ts#L187

I think this makes sense for most cases, but I personally would not mind having a longer timeout. The code it generates for me is worth it.

Currently I have to hack the source code to achieve this. Would it be possible to add this to the agent configuration? Such as:

```toml
## You can set the timeout for completion requests.
[completion.timeout]
auto = 5000 # ms, for auto completion when typing
manually = 30000 # ms, for manually triggered completion
healthy_limit = 4000 # ms <-- new
```

---
Please reply with a 👍 if you want this feature.

**Comment by wsxiaoys** [comment]
Since we've reintroduced the [server-side timeout option](https://github.com/TabbyML/tabby/commit/21ec60eddfc4d2948a771acf486691e04cafc5c1) (going to be released on v0.5), I agree that it should be fine to add back these options to `agent/config.toml`, especially since it's intended for guru users anyway. 

Given that you've already made it work locally, would you mind submitting a PR?

**Comment by icycodes** [comment]
> After 3 requests, considers the server unhealthy and then locks out further completions:
It only shows a warning for an unhealthy status, you should still be able to receive completions.
However, I agree that the threshold of 4 seconds for the warning should be configurable.

**Comment by f1zy2** [comment]
I was expecting the timeout to be the reason why I couldn't get any completion. But you say that we should still receive them besides that, right?

**Comment by f1zy2** [comment]
> > After 3 requests, considers the server unhealthy and then locks out further completions
> 
> It only shows a warning for an unhealthy status, you should still be able to receive completions. However, I agree that the threshold of 4 seconds for the warning should be configurable.

I was expecting the timeout to be the reason why I couldn't get any completion. But you say that we should still receive them besides that, right?

**Comment by icycodes** [comment]
> > > After 3 requests, considers the server unhealthy and then locks out further completions
> > 
> > 
> > It only shows a warning for an unhealthy status, you should still be able to receive completions. However, I agree that the threshold of 4 seconds for the warning should be configurable.
> 
> I was expecting the timeout to be the reason why I couldn't get any completion, given that I am running on a not so powerful laptop. But you say that we should still receive them besides the timeout, right?

Sorry for any confusion.   
The code you referred to **does not lock out further completions**. It simply sets a threshold to mark the server as unhealthy and displays a warning message.   
If a request times out due to either server or client timeout configuration, we cannot receive completions for that specific request. However, it does not block subsequent completion requests.

**Comment by icycodes** [comment]
> > Since we've reintroduced the [server-side timeout option](https://github.com/TabbyML/tabby/commit/21ec60eddfc4d2948a771acf486691e04cafc5c1) (going to be released on v0.5), I agree that it should be fine to add back these options to `agent/config.toml`, especially since it's intended for guru users anyway.
> 
> I'm glad to hear it.
> 
> > Given that you've already made it work locally, would you mind submitting a PR?
> 
> Unfortunately, I did it by patching the generated JS inside the VScode template, rather than editing the source properly.
> 
> That said, I did take a look. I am not very good at Typescript, but it does not look easy.
> 
> I figured out how to add the timeout value to the `AgentConfig`. The trouble is the line of code I highlighted. It is in a separate class which does nothing but calculate stats. It is a message that appears to be passed back and forth between the agent and the clients.
> 
> I could not figure out how to bring the "current agent" into that context, so I could call `getConfig()` on it and read the value I should be comparing against instead of the hard-coded `3000`. Am I missing something?

@jhwgh1968 Thank you! I will fix it in the next release.

**Comment by icycodes** [comment]
Released in extensions v1.1.0.
